### PR TITLE
[SPARK-31390][SQL][DOCS] Document Window Function in SQL Syntax Section

### PR DESCRIPTION
**Pull request by huaxingao**
### What changes were proposed in this pull request?
Document Window Function in SQL syntax

### Why are the changes needed?
Make SQL Reference complete

### Does this PR introduce any user-facing change?
Yes

Screen Shot 2020-04-16 at 9 13 34 PM

Screen Shot 2020-04-16 at 9 14 12 PM

Screen Shot 2020-04-16 at 9 14 45 PM

Screen Shot 2020-04-16 at 9 15 10 PM

Screen Shot 2020-04-16 at 9 15 25 PM

### How was this patch tested?
Manually build and check

### CODE REVIEW

**Comment by huaxingao** [comment]
cc @maropu

File: docs/sql-ref-syntax-qry-window.md [hunk 1]
```diff
@@ -19,4 +19,222 @@ license: |
...
 ---
 
-**This page is under construction**
+### Description
+
+Similarly to aggregate functions, window functions operate on a group of rows. However, unlike aggregate functions, window functions perform aggregation without reducing, calculating a return value for each row in the group. Window functions are useful for processing tasks such as calculating a moving average, computing a cumulative, or accessing the value of rows given the relative position of the current row. Spark SQL supports three types of window functions:
+  * Ranking Functions
+  * Analytic Functions
+  * Aggregate Functions
+
```
Comment (maropu) [inline docs/sql-ref-syntax-qry-window.md:28]: How about putting the `Syntax` section here before the `How to Use Window Functions` section?

File: docs/sql-ref-syntax-qry-window.md [hunk 2]
```diff
@@ -19,4 +19,222 @@ license: |
   limitations under the License.
 ---
 
-**This page is under construction**
+### Description
+
+Similarly to aggregate functions, window functions operate on a group of rows. However, unlike aggregate functions, window functions perform aggregation without reducing, calculating a return value for each row in the group. Window functions are useful for processing tasks such as calculating a moving average, computing a cumulative, or accessing the value of rows given the relative position of the current row. Spark SQL supports three types of window functions:
```
Comment (maropu) [inline docs/sql-ref-syntax-qry-window.md:24]: nit: add a single blank.

File: docs/sql-ref-syntax-qry-window.md [hunk 3]
```diff
@@ -19,4 +19,222 @@ license: |
...
+  * Ranking Functions
+  * Analytic Functions
+  * Aggregate Functions
+
+### How to Use Window Functions
+
+  * Mark a function as window function by using `over`.
+    - SQL: Add an OVER clause after the window function, e.g. rank ( ) OVER ( ... )
+    - DataFrame API: Call the window function's `over` method, e.g. rank ( ).over ( ... )
```
Comment (maropu) [inline docs/sql-ref-syntax-qry-window.md:33]: We need to describe the DataFrame APIs here in SQL references?

File: docs/sql-ref-syntax-qry-window.md [hunk 4]
```diff
@@ -19,4 +19,222 @@ license: |
...
 
-**This page is under construction**
+### Description
+
+Similarly to aggregate functions, window functions operate on a group of rows. However, unlike aggregate functions, window functions perform aggregation without reducing, calculating a return value for each row in the group. Window functions are useful for processing tasks such as calculating a moving average, computing a cumulative, or accessing the value of rows given the relative position of the current row. Spark SQL supports three types of window functions:
+  * Ranking Functions
+  * Analytic Functions
+  * Aggregate Functions
+
+### How to Use Window Functions
```
Comment (maropu) [inline docs/sql-ref-syntax-qry-window.md:29]: We need this section? The content looks duplicate with the `Parameters` section.

File: docs/sql-ref-syntax-qry-window.md [hunk 5]
```diff
@@ -19,4 +19,222 @@ license: |
...
+      <code>
+        { RANGE | ROWS } [ BETWEEN ]
+          UNBOUNDED { PRECEDING | FOLLOWING }
+          | CURRENT ROW
+          | boolean_expression { PRECEDING | FOLLOWING }
+      </code> <br><br>
+      <code>boolean_expression</code><br>
+      Specifies an expression with a return type of boolean.
+  </dd>
+</dl>
```
Comment (maropu) [inline docs/sql-ref-syntax-qry-window.md:118]: nit: add a single blank.

**Comment by maropu** [comment]
also cc: @viirya

File: docs/sql-ref-syntax-qry-window.md (hunk 2 above)
Comment (maropu) [inline docs/sql-ref-syntax-qry-window.md:24]: How about `Similarly to aggregate functions, window functions operate on a group of rows.` -> `A window function operates on a group of rows and this is comparable to aggregate functions. `?

Comment (maropu) [reply docs/sql-ref-syntax-qry-window.md:24]: `window functions perform aggregation without reducing, calculating a return value for each row in the group.` is not clear. This means `window functions do not compute a single aggregated value. Instead, they can generate multiple aggregated values for each group`?

File: docs/sql-ref-syntax-qry-window.md (hunk 3 above)
Comment (huaxingao) [reply docs/sql-ref-syntax-qry-window.md]: removed

File: docs/sql-ref-syntax-qry-window.md (hunk 4 above)
Comment (huaxingao) [reply docs/sql-ref-syntax-qry-window.md]: Removed

File: docs/sql-ref-syntax-qry-window.md (hunk 1 above)
Comment (huaxingao) [reply docs/sql-ref-syntax-qry-window.md:29]: The ```How to Use Window Functions``` is removed.

File: docs/sql-ref-syntax-qry-window.md [hunk 6]
```diff
@@ -19,4 +19,208 @@ license: |
   limitations under the License.
 ---
 
-**This page is under construction**
+### Description
+
+Similarly to aggregate functions, window functions operate on a group of rows. However, unlike aggregate functions, window functions perform aggregation without reducing, calculating an aggregated value for each row in the specified window. Window functions are useful for processing tasks such as calculating a moving average, computing a cumulative, or accessing the value of rows given the relative position of the current row. Spark SQL supports three types of window functions:
```
Comment (viirya) [inline docs/sql-ref-syntax-qry-window.md:24]: "without reducing"? Sounds confusing. How about "without reducing the number of rows"?

And "but calculating an aggregated value for each row in the specified window."

File: docs/sql-ref-syntax-qry-window.md [hunk 7]
```diff
@@ -19,4 +19,208 @@ license: |
   limitations under the License.
 ---
 
-**This page is under construction**
+### Description
+
+Window functions operate on a group of rows, referred to as a window, and calculate an aggregated value for each row based on the specified window. Window functions are useful for processing tasks such as calculating a moving average, computing a cumulative, or accessing the value of rows given the relative position of the current row. Spark SQL supports three types of window functions:
+
```
Comment (huaxingao) [inline docs/sql-ref-syntax-qry-window.md:25]: Does this look better now? @maropu @viirya

Comment (huaxingao) [reply docs/sql-ref-syntax-qry-window.md:25]: Also cc @srowen 
Please feel free to rephrase. Thanks!

Comment (srowen) [reply docs/sql-ref-syntax-qry-window.md:25]: computing a cumulative -> computing a cumulative _sum_ (or anything similar: average, statistic)

File: docs/sql-ref-syntax-qry-window.md [hunk 8]
```diff
@@ -19,4 +19,208 @@ license: |
...
+    </ul>
+  </dd>
+</dl>
+<dl>
+  <dt><code><em>partition_spec</em></code></dt>
+  <dd>
+    Specifies a comma separated list of key and value pairs for partitions.<br><br>
+    <b>Syntax:</b><br>
+      <code>
+        { PARTITION | DISTRIBUTE } BY partition_col_name  = partition_col_val ( [ , ... ] )
```
Comment (maropu) [inline docs/sql-ref-syntax-qry-window.md:75]: nit: I found the double spaces in this line.

File: docs/sql-ref-syntax-qry-window.md [hunk 9]
```diff
@@ -19,4 +19,208 @@ license: |
...
+    </ul>
+    <ul>
+      <li> Aggregate Functions </li>
+      <br>
+      <b>Syntax:</b>
+        <code>
+          MAX | MIN | COUNT | SUM | AVG | ...
+        </code>
+        <br>
+        Please refer <a href="api/sql/">here</a> for a complete list of Spark Aggregate Functions.
```
Comment (maropu) [inline docs/sql-ref-syntax-qry-window.md:65]: nit: `here` -> `the Built-in Function document`?

Comment (maropu) [reply docs/sql-ref-syntax-qry-window.md:65]: nit: `Spark Aggregate Functions.` -> `Spark aggregate functions.`?

File: docs/sql-ref-syntax-qry-window.md (hunk 7 above)
Comment (maropu) [reply docs/sql-ref-syntax-qry-window.md:25]: Thanks, it looks better. How about putting the last statement in a new line?; 
```
...the current row. 

Spark SQL supports three types of window functions:

  * Ranking Functions
  * Analytic Functions
  * Aggregate Functions
```

File: docs/sql-ref-syntax-qry-window.md (hunk 9 above)
Comment (huaxingao) [reply docs/sql-ref-syntax-qry-window.md]: I will put ```sql-ref-functions-builtin.html``` as the link for Built-in Function document. It's broken now but will work after your PR is in.

Comment (maropu) [reply docs/sql-ref-syntax-qry-window.md]: Ur, could you revert the link back? I'm currently not sure that my PR is target at 3.0.

File: docs/sql-ref-syntax-qry-window.md (hunk 7 above)
Comment (maropu) [reply docs/sql-ref-syntax-qry-window.md:26]: We need this list here? The `Syntax` section has the same list.

File: docs/sql-ref-syntax-qry-window.md [hunk 10]
```diff
@@ -19,4 +19,209 @@ license: |
...
+window_function OVER ( [ partition_spec ] order_spec [ window_frame ] )
+{% endhighlight %}
+
+### Parameters
+
+<dl>
+  <dt><code><em>window_function</em></code></dt>
+  <dd>
+    <ul>
+      <li> Ranking Functions </li>
```
Comment (maropu) [inline docs/sql-ref-syntax-qry-window.md:43]: nit: `      <li> Ranking Functions </li>` -> `      <li>Ranking Functions</li>`

File: docs/sql-ref-syntax-qry-window.md [hunk 11]
```diff
@@ -19,4 +19,209 @@ license: |
...
+    </ul>
+    <ul>
+      <li> Aggregate Functions </li>
+      <br>
+      <b>Syntax:</b>
+        <code>
+          MAX | MIN | COUNT | SUM | AVG | ...
+        </code>
+        <br>
+        Please refer to the <a href="sql-ref-functions-builtin.html">Built-in Function</a> document for a complete list of Spark aggregate functions.
```
Comment (maropu) [inline docs/sql-ref-syntax-qry-window.md:66]: nit: `Built-in Function` -> `Built-in Functions` by referring to the title in the doc: https://spark.apache.org/docs/latest/api/sql/index.html

File: docs/sql-ref-syntax-qry-window.md [hunk 12]
```diff
@@ -19,4 +19,209 @@ license: |
...
+      </code>
+  </dd>
+</dl>
+<dl>
+  <dt><code><em>order_spec</em></code></dt>
+  <dd>
+    Specifies an ordering of the rows.<br><br>
+    <b>Syntax:</b><br>
+      <code>
+        { ORDER | SORT } BY { expression [ ASC | DESC ] [ NULLS { FIRST | LAST } ] [ , ... ] }
```
Comment (maropu) [inline docs/sql-ref-syntax-qry-window.md:86]: Could you move `ORDER BY` and `PARTITION BY` caluses into the `Syntax` section like the Pg doc one? 
```
[ existing_window_name ]
[ PARTITION BY expression [, ...] ]
[ ORDER BY expression [ ASC | DESC | USING operator ] [ NULLS { FIRST | LAST } ] [, ...] ]
[ frame_clause ]
```
https://www.postgresql.org/docs/current/sql-select.html

File: docs/sql-ref-syntax-qry-window.md [hunk 13]
```diff
@@ -19,4 +19,209 @@ license: |
   limitations under the License.
 ---
 
-**This page is under construction**
+### Description
+
+Window functions operate on a group of rows, referred to as a window, and calculate an aggregated value for each row based on the specified window. Window functions are useful for processing tasks such as calculating a moving average, computing a cumulative statistic, or accessing the value of rows given the relative position of the current row.
```
Comment (viirya) [inline docs/sql-ref-syntax-qry-window.md:24]: "... calculate a return value for each row based on a group of rows"

File: docs/sql-ref-syntax-qry-window.md [hunk 14]
```diff
@@ -19,4 +19,192 @@ license: |
...
+  <dt><code><em>window_frame</em></code></dt>
+  <dd>
+    Specifies which row to start the window on and where to end it.<br><br>
+    <b>Syntax:</b><br>
+      <code>
+        { RANGE | ROWS } [ BETWEEN ]
+          UNBOUNDED { PRECEDING | FOLLOWING }
+          | CURRENT ROW
+          | boolean_expression { PRECEDING | FOLLOWING }
+      </code> <br><br>
```
Comment (maropu) [inline docs/sql-ref-syntax-qry-window.md:83]: I think we need to describe what these clauses (RANGE, ROWS, BETWEEN, ...) are.

File: docs/sql-ref-syntax-qry-window.md [hunk 15]
```diff
@@ -19,4 +19,194 @@ license: |
...
+        </code><br>
+        <code>offset:</code>specifies the <code>offset</code> from the position of the current row.
+      </ul>
+  </dd>
+</dl>
+
+### Examples
+
+{% highlight sql %}
+
```
Comment (maropu) [inline docs/sql-ref-syntax-qry-window.md:94]: nit: remove this blank.

**Review by maropu — APPROVED** [review]
Looks fine except for the existing two comments.
 - https://github.com/apache/spark/pull/28220#discussion_r409241050
 - https://github.com/apache/spark/pull/28220#discussion_r409915530

**Comment by maropu** [comment]
Could you update the screenshot in the description, too?

**Comment by huaxingao** [comment]
@maropu I have addressed the last two comments and updated the screenshots in description. Thanks for reviewing!

**Comment by huaxingao** [comment]
cc @srowen for final sign off.

File: docs/sql-ref-syntax-qry-window.md [hunk 16]
```diff
@@ -19,4 +19,188 @@ license: |
...
+  | Jane|  Marketing| 29000|29000|
+  | Jeff|  Marketing| 35000|29000|
+  | Fred|Engineering| 21000|21000|
+  |  Tom|Engineering| 23000|21000|
+  |Chloe|Engineering| 23000|21000|
+  | Paul|Engineering| 29000|21000|
+  +-----+-----------+------+-----+
+
+SELECT name, salary,
+    LAG(salary) OVER (PARTITION BY dept ORDER BY salary) as lag,
```
Comment (srowen) [inline docs/sql-ref-syntax-qry-window.md:185]: as -> AS 
but definitely don't change it just for that. Looks fine. I'll merge shortly

**Review by srowen — APPROVED** [review]

File: docs/sql-ref-syntax-qry-window.md [hunk 17]
```diff
@@ -19,4 +19,188 @@ license: |
...
+  |  Tom|Engineering| 23000|         2|
+  |Chloe|Engineering| 23000|         2|
+  | Paul|Engineering| 29000|         3|
+  |Helen|  Marketing| 29000|         1|
+  | Jane|  Marketing| 29000|         1|
+  | Jeff|  Marketing| 35000|         2|
+  +-----+-----------+------+----------+
+
+SELECT name, dept, age, CUME_DIST() OVER (PARTITION BY dept ORDER BY age
+    RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW) as cume_dist FROM employees;
```
Comment (maropu) [inline docs/sql-ref-syntax-qry-window.md:151]: nit: `as` -> `AS` here, too.

Comment (huaxingao) [reply docs/sql-ref-syntax-qry-window.md:151]: I will fix this

**Comment by maropu** [comment]
Thanks! Merged to master/3.0.

**Comment by huaxingao** [comment]
Thanks, all!

**Pull request closed without merging**